src/util/simplify_expr.cpp

```diff
@@ -236,22 +236,34 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
     return false;
   }
 
-  // eliminate casts to _Bool
-  if(expr_type.id()==ID_c_bool &&
-     op_type.id()!=ID_bool)
+  // circular casts through types shorter than `int`
+  const auto check_circular_cast=[&](const typet &type)
   {
-    // casts from boolean to a signed int and back:
-    // (boolean)(int)boolean -> boolean
-    if(expr.op0().id()==ID_typecast && op_type.id()==ID_signedbv)
+    if(expr_type==type)
     {
-      const auto &typecast=to_typecast_expr(expr.op0());
-      if(typecast.op().type().id()==ID_c_bool)
+      const auto &typecast=expr_dynamic_cast<typecast_exprt>(expr.op0());
```

Comment:
I come you know for a fact that the first operand is also a type cast?

Reply:
I see this in guaranteed by the check in the if, which isn't obvious. I suggest making a method that takes the `typecast_exprt` as an input and returns the modified `exprt`?

Reply:
Added a comment explaining how we know the `expr_dynamic_cast` is non-throwing